pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -220,12 +264,153 @@ func (r *ReconcileNodeMaintenance) stopNodeMaintenance(nodeName string) error {
+				return reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+
+func makeInt32(val int32) (*int32) {
+	tmpVal := val
+	return &tmpVal
+}
+
+func makeTimeNow() (*metav1.MicroTime) {
+
+	timeNow := metav1.MicroTime{Time: time.Now()}
+	return &timeNow
+}
+
+
+func (r *ReconcileNodeMaintenance) createOrGetExistingLease(node *corev1.Node, durationInSeconds int32) (*coordv1beta1.Lease, bool, error) {
+	nodeName := node.ObjectMeta.Name
+
+	holderIdentity := LeaseHolderIdentity
+
+	owner := &metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       nodeName,
+		UID:        node.ObjectMeta.UID,
+	}
+
+	microTimeNow :=  makeTimeNow()
+
+	lease := &coordv1beta1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            nodeName,
+			Namespace:       LeaseNamespace,
+			OwnerReferences: []metav1.OwnerReference{*owner},
+		},
+		Spec: coordv1beta1.LeaseSpec{
+			HolderIdentity:       &holderIdentity,
+			LeaseDurationSeconds: makeInt32(durationInSeconds),
+			AcquireTime:          microTimeNow,
+			RenewTime:            microTimeNow,
+			LeaseTransitions:     makeInt32(1),
+		},
+	}
+
+	if err := r.client.Create(context.TODO(), lease); err != nil {
+		if errors.IsAlreadyExists(err) {
+
+				nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+
+				if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+					return  nil, false, err
+				}
+				return lease, true, nil
+		}
+		return nil, false, err
+	}
+	return lease, false,  nil
+}
+
+func (r *ReconcileNodeMaintenance) updateLease(node *corev1.Node, lease *coordv1beta1.Lease, durationInSeconds int32) (*coordv1beta1.Lease, error) {
+
+	nodeName := node.ObjectMeta.Name
+
+	holderIdentity := LeaseHolderIdentity
+
+	owner := &metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       nodeName,
+		UID:        node.ObjectMeta.UID,
+	}
+
+	lease.ObjectMeta.Name = nodeName
+	lease.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*owner}
+	lease.Spec.HolderIdentity = &holderIdentity
+
+	timeNow := makeTimeNow()
+	lease.Spec.AcquireTime = timeNow
+	lease.Spec.RenewTime = timeNow
+
+	if lease.Spec.LeaseTransitions != nil {
+		*lease.Spec.LeaseTransitions += int32(1)
+	} else {
+		lease.Spec.LeaseTransitions = makeInt32(1)
+	}
+
+	lease.Spec.LeaseDurationSeconds = makeInt32(durationInSeconds)
+
+	if err := r.client.Update(context.TODO(), lease); err != nil {
+		log.Errorf("Failed to update the lease. node %s error: %v", nodeName, err)
+		return lease, err
+	}
+	log.Debugf("lease updated: duration: %d sec", durationInSeconds)
+
+	return lease, nil
+}
+
 func (r *ReconcileNodeMaintenance) fetchNode(nodeName string) (*corev1.Node, error) {
-	node := &corev1.Node{}
-	err := r.client.Get(context.TODO(), types.NamespacedName{Name: nodeName}, node)
+	node, err := r.drainer.Client.CoreV1().Nodes().Get( nodeName, metav1.GetOptions{})
```

Comment:
Reason?

Reply:
I had to use the clientset fake client for the unit test, instead of using the controller-runtime fake client. the controller-runtime fake client returned a node with an empty objectMeta field, so that I couldn't updateLease didn't. work. In update lease I probably have to add a check for a nil value before accessing the ObjectMeta field of the node. will add.